src/utils.c

```diff
@@ -110,30 +117,221 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+      else if ((!binverse && lcols[i]) || (binverse && !lcols[i]))
+        icols[ntrue++] = i+1;
     }
+    SETLENGTH(ricols, ntrue);
   } else {
-    error(_("argument specifying columns must be character or numeric"));
+    error(_("argument specifying columns must be character, numeric or logical"));
   }
   if (LOGICAL(check_dups)[0] && any_duplicated(ricols, FALSE))
     error(_("argument specifying columns specify duplicated column(s)"));
   UNPROTECT(protecti);
   return ricols;
 }
 
+/* replace_dot_alias
+ * we don't just simply alias .=list because i) list is a primitive (faster to iterate) and ii) we test for use
+ * of "list" in several places so it saves having to remember to write "." || "list" in those places
+ * replace_dot_aliasR wrapper should be used to prevent update of input
+ */
+SEXP replace_dot_alias(SEXP x) {
+  if (isLanguage(x) && !isFunction(CAR(x))) {
+    if (CAR(x) == sym_bquote) // handling `.` inside bquote, #1912
+      return x;
+    if (CAR(x) == sym_dot)
+      SETCAR(x, sym_list);
+    for (SEXP t=x; t!=R_NilValue; t=CDR(t)) {
+      if (!isNull(CADR(t)))
+        SETCADR(t, replace_dot_alias(CADR(t)));
+    }
+  }
+  return x;
+}
+SEXP replace_dot_aliasR(SEXP x) {
+  return replace_dot_alias(MAYBE_REFERENCED(x) ? duplicate(x) : x);
+}
+
+/* range_int
+ * used in exprCols(v4:v2) to produce 4:2 sequence
+ */
+SEXP range_int(int from, int to) {
+  if (from==NA_INTEGER || to==NA_INTEGER)
+    error(_("internal error: NA was supplied to generate integer range")); // # nocov
+  int len = to - from;
+  int ord = len>=0 ? 1 : -1;
+  len = len * ord;
+  SEXP ans = PROTECT(allocVector(INTSXP, len+1));
+  int *ansp = INTEGER(ans);
+  for (int i=0; i<=len; i++)
+    ansp[i] = from+i*ord;
+  UNPROTECT(1);
+  return ans;
+}
+
+/* funCols
+ * helper that calls lapply(DT, fun) to produce logical vector
+ */
+SEXP funCols(SEXP x, SEXP fun, SEXP rho) {
+  SEXP lfn = PROTECT(LCONS(sym_lapply, LCONS(x, LCONS(fun, R_NilValue))));
+  SEXP lst = PROTECT(eval(lfn, rho));
+  SEXP log = PROTECT(allocVector(LGLSXP, length(x)));
+  int *logp = LOGICAL(log);
+  for (int i=0; i<length(lst); ++i) {
+    SEXP lsti = VECTOR_ELT(lst, i);
+    if (length(lsti)!=1 || !isLogical(lsti) || LOGICAL(lsti)[0]==NA_LOGICAL)
+      error("When argument for columns selection is a function, it is applied to each column; the output of this function must be a non-missing boolean scalar signalling inclusion/exclusion of the column. However, these conditions were not met at least for: %s", CHAR(STRING_ELT(getAttrib(x, R_NamesSymbol), i)));
+    logp[i] = LOGICAL(lsti)[0];
+  }
+  UNPROTECT(3);
+  return log;
+}
+
+/* exprCols
+ * turns NSE expression into columns int
+ * helper for j and .SDcols
+ * handles v1, "v1", v1:v3, !"v1", -"v1", paste0("v",1:3)
+ * does not evaluate only for `symbol:symbol` expression, or `symbol` when "symbol" is existing column
+ */
+SEXP exprCols(SEXP x, SEXP expr, SEXP mode, /*SEXP with, */SEXP rho) {
```

Comment:
The current .SDcols, when inverse is TRUE, also looks at what ```bynames``` are available. In my R translation, I included a ```bynames``` argument that by default was ```NULL```.

Reply:
I would prefer to keep interaction with `bynames` as a post processing of result of this function